fix: Curve trade mainnet pools

adamazad opened the pull request:
# Summary

- Fixes estimations issues with Curve on Ethereum.
- Refactors pools ABI readability.
- Refactors tests for readability.
- Adds `pool` property to `CurveTrade`. The `pool` property is the Contract instance.
- Adds simulation tests for Curve swap on Ethereum due to high gas price. Arbitrum and GC tests have already passed.

adamazad commented on test/curve-trade.test.ts:
```diff
@@ -399,31 +564,57 @@ describe('CurveTrade', () => {
+            .sendTransaction(swapTransaction as any)
+            .then((tx) => tx.wait())
+            .catch(console.log)
+
+          console.log(exchangeTx)
+          // const logs = await unlockedAccountSigner.provider.getTransaction(exchangeTx.transactionHash)
+          // console.log(logs)
+
+          await new Promise((resolve) => setTimeout(resolve, 50000))
+
+          // expect(exchangeTx).toBeDefined()
```
```suggestion
```

berteotti commented on src/entities/trades/curve/CurveTrade.ts:
```diff
@@ -182,7 +190,7 @@ export class CurveTrade extends Trade {
     const amountInBN = parseUnits(currencyAmountIn.toSignificant(), tokenIn.decimals)
 
     // Determine if user has sent ETH
```
I think this comment is outdated

berteotti commented on src/entities/trades/curve/CurveTrade.ts:
```diff
@@ -282,7 +295,8 @@ export class CurveTrade extends Trade {
     const estimatedAmountOutPerPool = await Promise.all(
       routablePools.map(async (pool) => {
-        const poolContract = new Contract(pool.swapAddress, pool.abi as any, provider)
+        console.log(pool)
```
This `console.log` shouldn't exist

berteotti commented on src/entities/trades/curve/CurveTrade.ts:
```diff
@@ -301,14 +315,29 @@ export class CurveTrade extends Trade {
 
         // Debug
-        debug('Curve::GetQuote | Fetching estimated output', pool.swapAddress, dyMethodSignature, dyMethodParams)
+        console.log({
```
This `console.log` shouldn't exist

berteotti commented on src/entities/trades/curve/CurveTrade.ts:
```diff
@@ -301,14 +315,29 @@ export class CurveTrade extends Trade {
-        } catch (e) {
-          console.log(e)
+        } catch (error) {
+          console.log(`CurveTrade error: failed to fetch estimated out from `, {
```
Should this be a `console.error` instead of `console.log`?

berteotti commented on src/entities/trades/curve/CurveTrade.ts:
```diff
@@ -350,23 +379,40 @@ export class CurveTrade extends Trade {
+        return signature.startsWith('exchange(')
+      }) || 'exchange'
+
+    // console.log('poolContract.interface', poolContract.interface.getFunction(exchangeSignature))
```
This can be deleted

berteotti commented on src/entities/trades/curve/CurveTrade.ts:
```diff
@@ -350,23 +379,40 @@ export class CurveTrade extends Trade {
       debug(e)
     }
 
+    console.log({ tokens: pool.tokens, tokenIn })
```
This can be deleted

berteotti commented on src/entities/trades/curve/CurveTrade.ts:
```diff
@@ -350,23 +379,40 @@ export class CurveTrade extends Trade {
-      // debug(`Signature ${exchangeSignature} not found`)
-      // debug(poolContract.functions)
-      exchangeSignature = 'exchange(int128,int128,uint256,uint256)'
+    console.log(poolContract.functions)
```
This can be deleted

berteotti commented on src/entities/trades/curve/contracts/pools.ts:
```diff
@@ -0,0 +1,88 @@
+  }
+
+  const provider = getProvider(chainId)
+  // const multicallProvider = new MulticallProvider(provider, chainId)
+  // // // UniswapV3 multicall contract
+  // const multicall2Contract = new Contract('0x5ba1e12693dc8f9c48aad8770482f4739beed696', MULTICALL2_ABI, provider)
```
This can be deleted

berteotti commented on src/entities/trades/curve/pools/pools.ts:
```diff
@@ -0,0 +1,503 @@
+    abi: CURVE_3POOL_ABI,
+    isMeta: false,
+    tokens: [TOKENS_ARBITRUM_ONE.usdc, TOKENS_ARBITRUM_ONE.usdt],
+    // underlyingTokens: [TOKENS_ARBITRUM_ONE.usdc, TOKENS_ARBITRUM_ONE.usdt],
```
This can be deleted

berteotti commented on src/entities/trades/curve/tokens/tokens.ts:
```diff
@@ -0,0 +1,592 @@
+  },
+}
+
+// type StyleSet<T extends string> = {
+//   [k in T]: CurveToken
+// }
```
This can be deleted

adamazad replied:
Addressed in [59c2ccc](https://github.com/levelkdev/swapr-sdk/pull/71/commits/59c2ccc7ad2365df40c53cccdeab3949ccd8a83e)

berteotti commented on test/curve-trade.test.ts:
```diff
@@ -179,17 +148,17 @@ describe('CurveTrade', () => {
-  /*
-  describe.skip('Ethereum', () => {
+  describe('Ethereum', () => {
+    //
```
This can be deleted

berteotti commented on test/curve-trade.test.ts:
```diff
@@ -285,16 +331,148 @@ describe('CurveTrade', () => {
+      expect(swapTransaction?.value?.toString()).toEqual('0')
+      expect(swapTransaction?.to?.toLowerCase()).toEqual(CURVE_ROUTER_ADDRESS.toLowerCase())
+    })
+    test.skip('Should find a route from 1000 USDC to WETH via Curve Smart Router', async () => {
```
Should this test be skipped?

adamazad replied:
There's no route at the moment. Curve smart router doesn't always return a trade path. I skipped them for now.

adamazad commented on src/entities/trades/curve/pools/pools.ts:
```diff
@@ -0,0 +1,503 @@
+ */
+export const POOLS_XDAI: CurvePool[] = [
+  {
+    id: '3pool',
+    name: '3Pool',
+    abi: CURVE_3POOL_ABI,
+    address: '0x7f90122BF0700F9E7e1F688fe926940E8839F353',
+    isMeta: false,
+    // Order is crucial
+    tokens: [TOKENS_XDAI.wxdai, TOKENS_XDAI.usdc, TOKENS_XDAI.usdt],
+  },
+]
+
+export const POOLS_ARBITRUM_ONE: CurvePool[] = [
+  {
+    id: '2pool',
+    name: '2pool',
```
I got these from the official SDK. Curve is not standard.

adamazad commented:
@berteotti addressed your comments in [35233fd](https://github.com/levelkdev/swapr-sdk/pull/71/commits/35233fd3a308ae3c166df249dcfe487cad424736). Thanks!
